mg/builtins_test.go

```diff
@@ -0,0 +1,251 @@
+package mg_test
+
+import (
+	"bytes"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+
+	"margo.sh/mg"
+)
+
+func TestBultinCmdList_Lookup(t *testing.T) {
+	t.Parallel()
+	exec := func() mg.BultinCmd {
+		r, _ := mg.Builtins.Commands().Lookup(".exec")
+		return r
+	}()
+	item := mg.BultinCmd{
+		Name: "this name",
+		Desc: "description",
+		Run:  func(*mg.BultinCmdCtx) *mg.State { return nil },
+	}
+	tcs := []struct {
+		name      string
+		bcl       mg.BultinCmdList
+		input     string
+		wantCmd   mg.BultinCmd
+		wantFound bool
+	}{
+		{"empty cmd list", mg.BultinCmdList{}, "nothing to find", exec, false},
+		{"not found", mg.BultinCmdList{item}, "not found", exec, false},
+		{"found", mg.BultinCmdList{item}, item.Name, item, true},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			gotCmd, gotFound := tc.bcl.Lookup(tc.input)
+			// there is no way to compare functions, therefore we just check the names.
+			if gotCmd.Name != tc.wantCmd.Name {
+				t.Errorf("Lookup(): gotCmd = (%v); want (%v)", gotCmd, tc.wantCmd)
+			}
+			if gotFound != tc.wantFound {
+				t.Errorf("Lookup(): gotFound = (%v); want (%v)", gotFound, tc.wantFound)
+			}
+		})
+	}
+}
+
+// tests when the Args is empty, it should pick up the available BuiltinCmd(s).
+func TestTypeCmdEmptyArgs(t *testing.T) {
+	t.Parallel()
+	item1 := mg.BultinCmd{Name: "this name", Desc: "this description"}
+	item2 := mg.BultinCmd{Name: "another one", Desc: "should appear too"}
+	buf := new(bytes.Buffer)
+	input := &mg.BultinCmdCtx{
+		Ctx: &mg.Ctx{
+			State: &mg.State{
+				BuiltinCmds: mg.BultinCmdList{item1, item2},
+			},
+		},
+		Output: &mg.CmdOutputWriter{
+			Writer:   buf,
+			Dispatch: nil,
+		},
+	}
+
+	if got := mg.TypeCmd(input); !reflect.DeepEqual(got, input.State) {
+		t.Errorf("TypeCmd() = %v, want %v", got, input.State)
+	}
+	out := buf.String()
+	for _, item := range []mg.BultinCmd{item1, item2} {
+		if !strings.Contains(out, item.Name) {
+			t.Errorf("buf.String() = (%s); want (%s) in it", out, item.Name)
+		}
+		if !strings.Contains(out, item.Desc) {
+			t.Errorf("buf.String() = (%s); want (%s) in it", out, item.Desc)
+		}
+	}
+}
+
+func buildBultinCmdCtx(cmds mg.BultinCmdList, args []string, envMap mg.EnvMap, buf io.Writer) *mg.BultinCmdCtx {
+	ctx := &mg.Ctx{
```

Comment:
It's safest to use the `newXXX()` functions when available. They often make sure things are initialised correctly.
e.g. `newCtx` initialises things like `Store` because all the code assumes that if you give it a `*Ctx` then it can access all the fields.

Reply:
Good point. I will export a function to create a new ctx. I think you're already out of sync with the repo, but it shouldn't cause any merge conflicts

Reply:
I just pushed it. Use it like

	ag, _ := mg.NewAgent(mg.AgentConfig{})
	ctx, done := ag.NewCtx(nil)
	defer close(done)

Reply:
BTW, if I don't repond the some comments, ping me. This github ui is hard to follow and I think I might be missing some of your comments.

Reply:
:+1: